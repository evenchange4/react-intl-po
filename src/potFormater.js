/**
 * Ensure that multi-line strings are properly commented out
 *
 * For instance:
 * This is\nmy multi-line\ncomment
 *
 * should be escaped as :
 * #. This is
 * #. my multi-line
 * #. comment
 *
 * @param {String} commentPrefix
 * @param {String} rawComment
 * @returns {String}
 *
 * @author Guillaume Boddaert
 */
const potCommentMultiLineWrapper = (commentPrefix, rawComment) => {
  const comments = rawComment.split('\n');
  return comments.reduce((a, b) => `${a}${commentPrefix} ${b}\n`, '');
};


/**
 * Formatting POT comments
 * @param {Object[]} messageList
 * @return {String}
 *
 * example: see tests
 *
 * @author Michael Hsu
 * @author Guillaume Boddaert
 */
const potCommentsFormater = messageList =>
  messageList.reduce((acc, { filename, id, description, defaultMessage, mappedBy }) => {
    let out = acc;
    out += potCommentMultiLineWrapper('#:', filename);
    if (description) {
      out += potCommentMultiLineWrapper('#.', `[${id}] - ${description}`);
    } else {
      out += potCommentMultiLineWrapper('#.', `[${id}]`);
    }
    if (mappedBy && mappedBy !== 'defaultMessage') {
      out += potCommentMultiLineWrapper('#.', `defaultMessage is:\n${defaultMessage}`);
    }

    return out;
  }, '');

/**
 * Formatting POT comments
 * @param {Object} messageObject
 * @return {String}
 *
 * example: see tests
 *
 * @author Michael Hsu
 */

const potFormater = messageObject =>
  Object.keys(messageObject) // return array of id
    .sort()
    .map(id => `${potCommentsFormater(messageObject[id])}msgid "${id}"\nmsgstr ""\n`)
    .join('\n');

export default potFormater;
