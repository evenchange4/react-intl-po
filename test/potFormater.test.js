import test from 'ava';
import potFormater from '../src/potFormater';

test('should return a function', (t) => {
  t.is(typeof potFormater, 'function');
});

test('should return pot formatted string', (t) => {
  t.is(
    potFormater({
      'Go to MCS website': [
        {
          id: 'App.errorButton',
          description: 'Click error Button',
          defaultMessage: 'Go to MCS website',
          filename: './messages/src/containers/App/App.json',
        },
        {
          id: 'NotFound.errorButton',
          description: 'Click error Button',
          defaultMessage: 'Go to MCS website',
          filename: './messages/src/containers/NotFound/messages.json',
        },
      ],
    }),
    '#: ./messages/src/containers/App/App.json\n' +
    '#. [App.errorButton] - Click error Button\n' +
    '#: ./messages/src/containers/NotFound/messages.json\n' +
    '#. [NotFound.errorButton] - Click error Button\n' +
    'msgid "Go to MCS website"\n' +
    'msgstr ""\n',
  );
});

test('should return pot formatted string, with explicit mapping specified and not defaultMessage', (t) => {
  t.is(
    potFormater({
      'NotFound.errorButton': [
        {
          id: 'NotFound.errorButton',
          description: 'Click error Button',
          defaultMessage: 'Go to MCS website',
          filename: './messages/src/containers/NotFound/messages.json',
          mappedBy: 'id',
        },
      ],
    }),
    '#: ./messages/src/containers/NotFound/messages.json\n' +
    '#. [NotFound.errorButton] - Click error Button\n' +
    '#. defaultMessage is:\n' +
    '#. Go to MCS website\n' +
    'msgid "NotFound.errorButton"\n' +
    'msgstr ""\n',
  );
});

test('should return pot formatted string, with null or undefined description', (t) => {
  t.is(
    potFormater({
      'Go to MCS website': [
        {
          id: 'NotFound.errorButton',
          defaultMessage: 'Go to MCS website',
          filename: './messages/src/containers/NotFound/messages.json',
          mappedBy: 'defaultMessage',
        },
      ],
    }),
    '#: ./messages/src/containers/NotFound/messages.json\n' +
    '#. [NotFound.errorButton]\n' +
    'msgid "Go to MCS website"\n' +
    'msgstr ""\n',
  );
});


test('should return pot formatted string, with multi line values', (t) => {
  t.is(
    potFormater({
      'NotFound.errorButton': [
        {
          id: 'NotFound.errorButton',
          description: 'My description\nis\nquite\nlong.',
          defaultMessage: 'This is\nmultiline',
          filename: './messages/src/containers/NotFound/messages.json',
          mappedBy: 'id',
        },
      ],
    }),
    '#: ./messages/src/containers/NotFound/messages.json\n' +
    '#. [NotFound.errorButton] - My description\n' +
    '#. is\n' +
    '#. quite\n' +
    '#. long.\n' +
    '#. defaultMessage is:\n' +
    '#. This is\n' +
    '#. multiline\n' +
    'msgid "NotFound.errorButton"\n' +
    'msgstr ""\n',
  );
});
