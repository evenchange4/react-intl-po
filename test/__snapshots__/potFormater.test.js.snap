// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return pot formatted string 1`] = `
#: ./messages/src/containers/App/App.json
#. [App.errorButton] - Click error Button
#. defaultMessage is:
#. Go to MCS website
#: ./messages/src/containers/NotFound/messages.json
#. [NotFound.errorButton] - Click error Button
#. defaultMessage is:
#. Go to MCS website
msgid "Go to MCS website"
msgstr ""

`;

exports[`should return pot formatted string, with multi line values 1`] = `
#: ./messages/src/containers/NotFound/messages.json
#. [NotFound.errorButton] - My description
#. is
#. quite
#. long.
#. defaultMessage is:
#. This is
#. multiline
msgid "NotFound.errorButton"
msgstr ""

`;

exports[`should return pot formatted string, with null or undefined description 1`] = `
#: ./messages/src/containers/NotFound/messages.json
#. [NotFound.errorButton]
#. defaultMessage is:
#. Go to MCS website
msgid "Go to MCS website"
msgstr ""

`;
